### DISCUSSION TITLE
Fetch all data from Netsuite

### DISCUSSION THREAD

**Discussion by abirahsan007** [netsuitephp/netsuite-php#123] [Q&A] [answered]
So I have Netsuite Login creds , Integration Details and Access Token for one of m site. How can i fetch all data with API in a PHP file?

**Comment by neclimdul** [comment, marked as answer]
"All" is pretty complicated. The schema is massive so depending on what you're looking for its going to have a slightly different answer. [The schema browser](https://system.netsuite.com/help/helpcenter/en_US/srbrowser/Browser2022_1/schema/record/account.html) is probably going to be your friend. Everything in that schema should alight to something in the toolkit.

If you're truly looking for "everything" I think you're going to be using [search](https://github.com/netsuitephp/netsuite-php/blob/master/src/NetSuiteService.php#L80) or [getAll](https://github.com/netsuitephp/netsuite-php/blob/master/src/NetSuiteService.php#L190). 

I don't think I've ever really used getAll because it wasn't paged and so the result are unwieldy. Probably useful on small list but Netsuite can be slow and soap is verbose so getting the response to 100s of thousands of contacts can time out and be 100s of megabytes or even gigabytes to load into memory.

To run a search, first, you'll need to start with this to setup the service object:
https://github.com/netsuitephp/netsuite-php#quickstart

Then, depending on what you're looking for you're going to create a search request and pass it to the search method on the service. There are lots of record types as mentioned and in the case of some like custom records you'll need multiple passes. 

```php
$id = 1; // custom entity type id. There may be a lot, hopefully your integration docs specify which ones.
$search = new CustomRecordSearchBasic();
$search->recType = new RecordRef();
$search->recType->internalId = (string) $id;
$search->recType->type = 'customRecord';
$request = new SearchRequest();
$request->searchRecord = $search;
$searchResponse = $service->search($request);
```

Even this has some complications though. As mentioned, search is paged. However unlike what you might see in a rest service, the paging isn't handled by the same method. The initial search with `search` creates a session and subsequent paged requests happens on `searchMore`. Its a bit complicated but you can google answers to get the idea. This project might be useful for taking care are of it or explaining it better as well. https://github.com/neclimdul/netsuite-search-iterator

Hope that's what you're looking for. Good luck!
